### PR TITLE
IBC design rationale & architecture document

### LINKED ISSUE #49: IBC definitions
Just noting for clarity so we avoid using the same term to mean different things in different places.

**Ledger / Chain**

A *ledger*, or *chain*, is a distributed ledger (or "blockchain", although a strict chain of blocks may not be required) implementing part or all of the IBC specification as a component or module within its state machine.

**Handler**

An IBC *handler* is the module or subcomponent within the state machine of a ledger responsible for implementing the IBC specification by "handling" datagrams, performing the appropriate checks, proof verifications,  and/or storage alterations, and routing valid packets to other parts of the state machine, as defined by the application-layer semantics.

**Datagram**

A *datagram* is an opaque data blob transmitted over some network, and handled by the top-level IBC handler implemented in the ledger's state machine. In some implementations, the datagram may be a field in a ledger-specific transaction or message data structure which also contains other information (e.g. a fee for spam prevention, nonce for replay prevention, type identifier to route to the IBC handler, etc.)

**Connection**

A *connection* is a persistent data structure on a particular ledger that contains information about the consensus state of another ledger. Updates to that consensus state change the state of the connection.

**Channel**

A *channel* is a persistent data structure on a particular ledger that contains metadata to facilitate packet ordering, exactly-once delivery, and replay prevention. Often it may make sense to reason about a "channel" across two chains (reasoning about subsets of state on both ledgers).

**Packet** 

A *packet* is a particular data structure with sequence-related metadata (defined by the IBC specification) and an opaque value field (with semantics defined by the application layer).

### LINKED ISSUE #51: IBC statement of purpose
Somewhere we should have a document outlining, at a high level, the purpose, design goals, and scope of IBC, and linking to the different subcomponents. This document should be designed to be read by someone who has some background in blockchains but none with Tendermint/Cosmos or IBC.

This isn't a technical specification, so I don't think it should go in this repository. Maybe it should be a page on the Cosmos or Interchain website - the closest kind of document to what I'm envisioning is the [What is Cosmos](https://cosmos.network/intro) page on the Cosmos Network website (which has a section on IBC, but not a comprehensive breakdown).

### PR DESCRIPTION
**Pull request by cwgoes**
Replaces https://github.com/cosmos/ics/pull/55
Closes #51 
Closes #49 
Ref https://github.com/cosmos/ics/issues/45 but that one can remain open

- Enumerates the high-level design goals into IBC
- Explains how they have led to the particular choices we have made
- Explains the broad architecture of IBC in plain English
- Provides high-level definitions for terminology used but not defined in individual ICSs

I expect this could benefit in the future from visual graphics as well.

### CODE REVIEW

**Comment by cwgoes** [comment]
cc @ancazamfir - I can't seem to request you as a reviewer, but I'd appreciate your feedback.

**Comment by cwgoes** [comment]
> First thoughts...for the architecture document, could we have maybe:
> 
>     * a reference architecture diagram showing the actor, chains, the different modules, channels and connection, etc
> 
>     * point the different ICS-es to the elements in the architecture diagram that they specify (could be multiple if we split a spec in multiple ICS-es e.g. for the purpose of ease of collaboration)
> 
>     * hight level explantation of the interactions between the different elements of the architecture in a plain English story.
> 
> 
> Currently this doc is a list of definitions and terms and IMHO not enough to give a good high level view of where each spec fits.

Yes, I think you're right - I'm not quite sure how best to diagram the system, but that would be ideal, I can play around. An architectural pointer to sub-standards seems very useful, and the text could at least explain the stages in an example IBC transaction transferring tokens between two chains.

File: docs/IBC_ARCHITECTURE.md [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+An *actor*, or a *user* (used interchangeably), is an entity interacting with the IBC protocol. An actor can be a human end-user, a module or smart contract running on a blockchain, or an off-chain relayer process capable of signing transactions.
+
+#### Chain
+
+A *chain*, or *ledger* (used interchangeably), is a distributed ledger (or "blockchain", although a strict chain of blocks may not be required) implementing part or all of the IBC specification as a component or module within its state machine.
+
+#### State Machine
+
+The *state machine* of a particular ledger is the set of rules which determines valid transactions and blocks based on the current state agreed upon by the consensus algorithm of the ledger.
```
[the remark this answers is not in the document]
Comment (cwgoes) [reply docs/IBC_ARCHITECTURE.md]: Applied

File: docs/IBC_ARCHITECTURE.md [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+A *channel* is a set of persistent data structures on particular ledgers (usually two) that contain metadata to facilitate packet ordering, exactly-once delivery, and replay prevention. Packets sent through a channel change its internal state.
+
+#### Packet
+
+A *packet* is a particular data structure with sequence-related metadata (defined by the IBC specification) and an opaque value field (with semantics defined by the application layer, e.g. token amount and denomination).
+
+#### Module
+
+A *module* is a subcomponent of the state machine of a particular blockchain which interacts with the IBC handler and alters state according to particular IBC packets sent or received (minting or burning tokens, for example).
```
[the remark this answers is not in the document]
Comment (cwgoes) [reply docs/IBC_ARCHITECTURE.md]: Sure, updated.

File: docs/IBC_ARCHITECTURE.md [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+A *consensus* algorithm is the protocol used by the set of processes operating a distributed ledger to come to agreement on the same state, generally under the presence of a bounded number of Byzantine faults.
+
+#### Handler
+
+An IBC *handler* is the module or subcomponent within the state machine of a ledger responsible for implementing the IBC specification by "handling" datagrams, performing the appropriate checks, proof verifications, and/or storage alterations, and routing valid packets to other parts of the state machine, as defined by the application-layer semantics.
+
+#### Datagram
+
+A *datagram* is an opaque data blob transmitted over some physical network, and handled by the top-level IBC handler implemented in the ledger's state machine. In some implementations, the datagram may be a field in a ledger-specific transaction or message data structure which also contains other information (e.g. a fee for spam prevention, nonce for replay prevention, type identifier to route to the IBC handler, etc.). All IBC subprotocols (such as opening a connection, creating a channel, sending a packet) are defined in terms of sets of datagrams and protocols for handling them.
```
[the remark this answers is not in the document]
Comment (cwgoes) [reply docs/IBC_ARCHITECTURE.md]: Binary blog - switched to "bytestring", is that better?

File: docs/WHY_IBC.md [hunk 4]
```diff
@@ -0,0 +1,45 @@
+## Why IBC?
```
[the remark this answers is not in the document]
Comment (cwgoes) [reply docs/WHY_IBC.md]: Because it looks too large on my monitor - but fair enough. Updated throughout.

File: docs/IBC_ARCHITECTURE.md [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+#### Module
+
+A *module* is a subcomponent of the state machine of a particular blockchain which interacts with the IBC handler and alters state according to particular IBC packets sent or received (minting or burning tokens, for example).
+
+### Auxiliary Terms
+
+#### Handshake
+
+A *handshake* is a particular class of subprotocol involving multiple datagrams, generally used to initialize some state on multiple chains.
```
[the remark this answers is not in the document]
Comment (cwgoes) [reply docs/IBC_ARCHITECTURE.md]: Conceivably yes but for now no. Clarified, hopefully.

File: docs/WHY_IBC.md [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+### Modularity
+
+IBC is designed to be a *modular* protocol. The protocol is constructed as a series of layered components with explicit security properties & requirements. Implementations of a component at a particular layer can vary (such as a different consensus algorithm or connection opening procedure) as long as they provide the requisite properties to the higher layers (such as finality, < 1/3 Byzantine safety, or embedded roots-of-trust on two chains). State machines need only understand compatible subsets of the IBC protocol (e.g. lite client verification algorithms for each other's consensus) in order to safely interact.
+
+### Locality
+
+IBC is designed to be a *local* protocol, meaning that only information about the two connected chains is necessary to reason about the security and correctness of a bidirectional IBC connection. Security requirements of the authentication primitives refer only to consensus algorithms and validator sets of the blockchains involved in the connection, and blockchains maintaining a set of IBC connections need only understand the state of the chains to which they are connected (no matter which other chains those chains are connected to).
```
[the remark this answers is not in the document]
Comment (cwgoes) [reply docs/WHY_IBC.md]: The IBC protocol, for now and for the immediate future, will be defined between only two chains - higher-level protocols could be used for multi-hop routing, but that can be built on top of the simpler two-chain core protocol.

I think I only say "multiple chains" when talking about a set of chains implementing the protocol - where is it unclear? I changed this sentence a bit, hopefully that helps.

File: docs/IBC_TERMINOLOGY.md [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+
+### Connection
+
+A *connection* is a set of persistent data structures on two chains that contain information about the consensus state of the other ledger in the connection. Updates to the consensus state of one chain changes the state of the connection object on the other chain.
+
+### Channel
+
+A *channel* is a set of persistent data structures on two chains that contain metadata to facilitate packet ordering, exactly-once delivery, and replay prevention. Packets sent through a channel change its internal state. Channels are associated with connections in a many-to-one relationship — a single connection can have any number of associated channnels, and all channels must have a single associated connection.
```
Comment (mossid) [inline docs/IBC_TERMINOLOGY.md]: Should we add a definition for a `Port`, `Channel Class`, `Channel Set` or something named like that? Multiple channels can exist on a single connection, but it will be defined as a pair of `(ChannelClassID, ConnectionID)`.

File: docs/IBC_ARCHITECTURE.md [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+
+> For definitions of terms used in IBC specifications, see [here](./IBC_TERMINOLOGY.md).
+
+This document outlines the architecture of the authentication, transport, and ordering layers of the IBC protocol stack.
+
+## What is IBC?
+
+IBC is an application-agnostic layered protocol stack for *inter-blockchain communication* which handles authentication, transport, and ordering of structured data packets relayed between two blockchains.
+
+## What is IBC not?
```
Comment (jackzampolin) [inline docs/IBC_ARCHITECTURE.md]: Love this section!

File: docs/IBC_ARCHITECTURE.md [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+
+Dataflow on chain *A*:
+
+Actor --> Module --> Handler -> Packet --> Channel --> Connection --> Consensus -->
+
+---
+
+Off-chain:
+
+--> Relayer -->
```
Comment (jackzampolin) [inline docs/IBC_ARCHITECTURE.md]: So every IBC connection needs a relayer?

Comment (zmanian) [reply docs/IBC_ARCHITECTURE.md]: Relayer is a role rather than specific entity.

The specific entity fulfilling the relayer role is adhoc.

File: docs/IBC_DESIGN_PHILOSOPHY.md [hunk 10]
```diff
@@ -0,0 +1,49 @@
+### IBC Design Philosophy
```
[the remark this answers is not in the document]
Comment (cwgoes) [reply docs/IBC_DESIGN_PHILOSOPHY.md]: Numbered and moved to `docs/ibc`

File: docs/IBC_ARCHITECTURE.md [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+
+A *datagram* is an opaque data blob transmitted over some physical network, and handled by the top-level IBC handler implemented in the ledger's state machine. In some implementations, the datagram may be a field in a ledger-specific transaction or message data structure which also contains other information (e.g. a fee for spam prevention, nonce for replay prevention, type identifier to route to the IBC handler, etc.). All IBC subprotocols (such as opening a connection, creating a channel, sending a packet) are defined in terms of sets of datagrams and protocols for handling them.
+
+#### Connection
+
+A *connection* is a set of persistent data structures on particular ledgers (usually two) that contain information about the consensus state of the other ledgers in the connection. Updates to their consensus states change the state of the connections.
+
+#### Channel
+
+A *channel* is a set of persistent data structures on particular ledgers (usually two) that contain metadata to facilitate packet ordering, exactly-once delivery, and replay prevention. Packets sent through a channel change its internal state.
```
[the remark this answers is not in the document]
Comment (cwgoes) [reply docs/IBC_ARCHITECTURE.md]: So clarified.

File: docs/IBC_ARCHITECTURE.md (hunk 9 above)
Comment (cwgoes) [reply docs/IBC_ARCHITECTURE.md]: Clarified this.

File: docs/IBC_TERMINOLOGY.md [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+
+A *header* is an update to the consensus state of a particular blockchain that can be verified in a well-defined fashion by a "light client" algorithm defined for that particular consensus algorithm.
+
+### Finality
+
+*Finality* is the guarantee provided by a consensus algorithm that a particular block will not be reverted, subject to certain assumptions about the behavior of the validator set. The IBC protocol requires finality.
+
+### Accumulator
+
+A cryptographic *accumulator* is a way to cheaply verify membership of a value in a set, where the set can be commited to with a short witness string. An *accumulator proof* refers to the proof structure which proves that a particular value is a member of a particular committed-to set.
```
[the remark this answers is not in the document]
Comment (cwgoes) [reply docs/IBC_TERMINOLOGY.md]: Thanks, fixed.

File: docs/IBC_ARCHITECTURE.md [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+
+1. On chain *A*
+    1. Actor (application-specific)
+    1. Module (application-specific)
+    1. Handler (parts defined in different ICSs)
+    1. Packet (defined in [ICS 5](../spec/ics-5-packet-semantics))
+    1. Channel (defined in [ICS 4](../spec/ics-4-channel-semantics))
+    1. Connection (defined in [ICS 3](../spec/ics-3-connection-semantics))
+    1. Consensus (defined in [ICS 2](../spec/ics-2-consensus-requirements))
+2. Off-chain
```
[the remark this answers is not in the document]
Comment (cwgoes) [reply docs/IBC_ARCHITECTURE.md]: I don't quite follow - they are symmetrical (except for the initial actor), the relayer is the "line of reflection" if you like, as soon as chain A's consensus commits a packet the relayer sends a transaction to chain B, then chain B's consensus commits the incoming packet.

File: docs/IBC_TERMINOLOGY.md (hunk 7 above)
Comment (cwgoes) [reply docs/IBC_TERMINOLOGY.md]: Identifier details are too low-level for this document IMO. I clarified the many-to-one relationship.

File: docs/WHY_IBC.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (cwgoes) [reply docs/WHY_IBC.md]: I think we would actually already support that sort of case, where a module can connect to another module over IBC, then allow that module to send it roots-of-trust for other connections to open - e.g. a nameservice might work like this.

I'm just talking about in terms of defining the connection / channel constructions.

File: docs/ibc/2_IBC_TERMINOLOGY.md [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+
+The *state machine* of a particular chain defines the structure of the state as well as the set of rules which determines valid transactions that trigger state-transitions based on the current state agreed upon by the consensus algorithm of the chain.
+
+### Consensus
+
+A *consensus* algorithm is the protocol used by the set of processes operating a distributed ledger to come to agreement on the same state, generally under the presence of a bounded number of Byzantine faults.
+
+### Consensus State
+
+The *consensus state*, or the *root-of-trust*, is the set of information about the state of a consensus algorithm required to verify proofs about the output of that consensus algorithm (e.g. accumulator roots in signed headers).
```
Comment (Liamsi) [inline docs/ibc/2_IBC_TERMINOLOGY.md]: Isn't "root of trust" a bit overloaded with a different meaning?

Comment (cwgoes) [reply docs/ibc/2_IBC_TERMINOLOGY.md]: I guess it is in Tendermint (at least in the codebase). Should we just use "root of trust" here?

File: docs/ibc/2_IBC_TERMINOLOGY.md [hunk 15]
```diff
@@ -0,0 +1,85 @@
...
+
+A *handshake* is a particular class of subprotocol involving multiple datagrams, generally used to initialize some common state on the two involved chains such as roots-of-trust for each others' consensus algorithms.
+
+### Trust
+
+To *trust* a blockchain or validator set means to expect that the validator set will behave in a particular way (such as < 1/3 Byzantine) relative to a well-defined consensus & state machine protocol.
+
+### Authorization
+
+*Authorization* refers to the protocols used to ensure that datagrams were in fact sent by a particular chain and associated state alterations committed by it.
```
[the remark this answers is not in the document]
Comment (cwgoes) [reply docs/ibc/2_IBC_TERMINOLOGY.md]: You are quite correct, I don't know why I wrote "Authorization". Fixed!

File: docs/ibc/3_IBC_ARCHITECTURE.md [hunk 16]
```diff
@@ -0,0 +1,69 @@
...
+    1. Consensus (defined in [ICS 2](../spec/ics-2-consensus-requirements))
+2. Off-chain
+    1. Relayer (defined in [ICS 18](../spec/ics-18-offchain-relayer))
+3. On chain *B*
+    1. Consensus (defined in [ICS 2](../spec/ics-2-consensus-requirements))
+    1. Connection (defined in [ICS 3](../spec/ics-3-connection-semantics))
+    1. Channel (defined in [ICS 4](../spec/ics-4-channel-semantics))
+    1. Packet (defined in [ICS 5](../spec/ics-5-packet-semantics))
+    1. Handler (parts defined in different ICSs)
+    1. Module (application-specific)
```
[the remark this answers is not in the document]
Comment (cwgoes) [reply docs/ibc/3_IBC_ARCHITECTURE.md]: They are. One (the former) is intended to become a visual diagram, and the latter is intended to remain a plain-English description. Still working on this part.

**Comment by cwgoes** [comment]
Merging this as a first version, following up with improvements inspired by the TCP RFC in another PR.

File: docs/ibc/1_IBC_DESIGN_PHILOSOPHY.md [hunk 17]
```diff
@@ -0,0 +1,49 @@
...
+
+### Locality of permissioning
+
+Actions in IBC — such as opening a connection, creating a channel, or sending a packet — are permissioned locally by the state machines and actors involved in a particular connection between two chains. Individual chains could choose to require approval from a permissioning mechanism (such as governance) for specific application-layer actions (such as delegated-security slashing), but for the base protocol, actions are permissionless (modulo gas & storage costs) — by default, connections can be opened, channels created, and packets sent without any approval process. Of course, users themselves must inspect the state & consensus of each IBC connection and decide whether it is safe to used (based e.g. on the roots-of-trust stored).
```
Comment (melekes) [inline docs/ibc/1_IBC_DESIGN_PHILOSOPHY.md]: > of course, users themselves must inspect the state & consensus of each IBC connection and decide whether it is safe to used

can this be done automatically?

Comment (cwgoes) [reply docs/ibc/1_IBC_DESIGN_PHILOSOPHY.md]: A client-side wallet could have some sort of safety heuristic, if that's what you mean.
